### DISCUSSION TITLE
Is it possible to boot a unified kernel image? (err, mkosi image)

### DISCUSSION THREAD

**Discussion by ajf** [ipxe/ipxe#632] [Q&A]
There's some tools that can generate a Unified Kernel Image (https://wiki.archlinux.org/title/Unified_kernel_image), but it doesn't seem like iPXE can directly boot such a file directly.  It would simplify our OS management a great deal if it's technically possible to boot such a thing within iPXE.

**Comment by mcb30** [comment]
That linked page states that the unified kernel image is a valid UEFI executable, in which case
iPXE will already be able to boot it.

**Comment by NiKiZe** [comment]
What makes you think iPXE can't boot it?
The only way to boot a Linux kernel from iPXE in EFI mode is to have the kernel compiled with EFI_STUB enabled.

Why Arch seems to call this "Unified kernel image" is however unclear.

**Reply by stappersg** [reply]
> Why Arch seems to call this "Unified kernel image" is however unclear.

My guess:  a mix of UEFI and kernel

My appeal to @ajf :
Please elaborate the problem that you are facing.
The _but it doesn't seem like iPXE can directly boot such a file directly_ is very vague.
I'm blunt enough to say _there was no attempt_ and I like to be proven wrong about that statement.

**Comment by ajf** [comment]
Oh wow, I totally misunderstood my own problem!

It's not a problem booting a unified kernel image, I'm actually failing at booting a raw disk image with a unified kernel image inside of it.  I didn't elaborate in my original post to let anyone know what I was trying to do.  I apologize for being so terse and wrong.

What I'm _actually_ trying to do:

Build a CI/CD pipeline for in-memory, read-only images that are booted from iPXE on UEFI machines (actual bare metal, not VMs).  Rather than invent my own build tools for making those images, I'm trying to use [mkosi](https://github.com/systemd/mkosi).

mkosi supports [multiple output formats](https://github.com/systemd/mkosi/blob/main/mkosi.md#supported-output-formats), and i'm trying to get one to boot directly from iPXE - right now I'm trying gpt_squashfs but I guess any would be useful.  Inside this image (if it's made bootable) is [sd-boot](https://www.freedesktop.org/software/systemd/man/systemd-boot.html) as well as an EFI partition with the unified kernel on it.

So the resulting image looks like this:

```
> file image.raw
image.raw: DOS/MBR boot sector; partition 1 : ID=0xee, start-CHS (0x0,0,2), end-CHS (0x3ff,255,63), startsector 1, 992687 sectors, extended partition table (last)
> sudo losetup -P /dev/loop13 image.raw
> sudo parted /dev/loop13
GNU Parted 3.4
Using /dev/loop13
Welcome to GNU Parted! Type 'help' to view a list of commands.
(parted) p
Model: Loopback device (loopback)
Disk /dev/loop13: 508MB
Sector size (logical/physical): 512B/512B
Partition Table: gpt
Disk Flags:

Number  Start   End    Size   File system  Name                  Flags
 1      20.5kB  268MB  268MB  fat32        ESP System Partition  boot, esp
 2      268MB   508MB  240MB               Root Partition

(parted)
```

When I boot this directly, I unsurprisingly got an exec format error.

I'd imagine trying to boot such a thing from iPXE when it's downloaded from http isn't currently possible?

**Reply by NiKiZe** [reply]
Sanboot will probably boot this fine, but remember that as soon as the kernel is loaded, any access to this in memory image will be lost, so why are you using the image instead of loading kernel and initrd from iPXE directly?

**Reply by ajf** [reply]
yeah, that's a good point.  The kernel would load and then will have to go figure out everything itself again anyway since there's nothing persistent and it's all in memory.

Seems like the "correct" behavior would be to boot the UEFI kernel image directly (after extracting it from the image.raw) via `chain` and tell dracut that the "live" root FS is the root partition so dracut would download it and pivot into the downloaded root.

I did some of that work in https://github.com/systemd/mkosi/issues/302#issuecomment-996472930 but wanted to see if we could boot it without splitting the artifacts apart.  Sounds like that's not necessary with sanboot, but has other show stoppers.

**Reply by ajf** [reply]
I guess to answer "why", the answer is so that when we publish artifacts we only have to publish and manage a single file with the complete OS in it.

**Reply by NiKiZe** [reply]
Then hopefully sanboot works for your case.
You should also be able to include initrd in kernel and that way only get one image as well.
